modules/transport/Transport.js

```diff
@@ -0,0 +1,239 @@
+import {
+    MESSAGE_TYPE_EVENT,
+    MESSAGE_TYPE_RESPONSE,
+    MESSAGE_TYPE_REQUEST
+} from './constants';
+
+/**
+ * Stores the currnet transport that have to be used.
+ */
+export default class Transport {
+    /**
+     * Creates new instance.
+     *
+     * @param {Object} options - Optional parameters for configuration of the
+     * transport.
+     */
+    constructor({ transport } = {}) {
+        this._requestID = 0;
+
+        this._responseHandlers = new Map();
+
+        this._listeners = new Map();
+
+        this._unprocessedMessages = new Set();
+
+        this.addListener = this.on;
+
+        if (transport) {
+            this.setTransport(transport);
+        }
+    }
+
+    /**
+     * Disposes the current transport.
+     *
+     * @returns {void}
+     */
+    _disposeTransport() {
+        if (this._transport) {
+            this._transport.dispose();
+            this._transport = null;
+        }
+    }
+
+    /**
+     * Handles incomming data from the transport.
+     *
+     * @param {Object} data - The data.
+     * @returns {void}
+     */
+    _onDataReceived(data) {
+        if (data.type === MESSAGE_TYPE_RESPONSE) {
+            const handler = this._responseHandlers.get(data.id);
+
+            if (handler) {
+                handler(data);
+                this._responseHandlers.delete(data.id);
+            }
+
+            return;
+        }
+
+        if (data.type === MESSAGE_TYPE_REQUEST) {
+            this.emit('request', data.data, (result, error) => {
+                this._transport.send({
+                    type: MESSAGE_TYPE_RESPONSE,
+                    error,
+                    id: data.id,
+                    result
+                });
+            });
+        } else {
+            this.emit('event', data.data);
+        }
+    }
+
+    /**
+     * Disposes the allocated resources.
+     *
+     * @returns {void}
+     */
+    dispose() {
+        this._responseHandlers.clear();
+        this._unprocessedMessages.clear();
+        this.removeAllListeners();
+        this._disposeTransport();
+    }
+
+    /**
+     * Calls each of the listeners registered for the event named eventName, in
+     * the order they were registered, passing the supplied arguments to each.
+     *
+     * @param {string} eventName -  The name of the event.
+     * @returns {boolean} True if the event had listeners, false otherwise.
+     */
+    emit(eventName, ...args) {
+        const listenersForEvent = this._listeners.get(eventName);
+        let isProcessed = false;
+
+        if (listenersForEvent && listenersForEvent.size) {
+            listenersForEvent.forEach(listener => {
+                isProcessed = listener(...args) || isProcessed;
+            });
+        }
+
+        if (!isProcessed) {
+            this._unprocessedMessages.add(args);
+        }
+
+        return isProcessed;
```

Comment:
I believe you we are not using it! I tried to implement this interface https://nodejs.org/api/events.html .

In our case I decided that it will be maybe useful to return whether the event was processed or not but the comment was not good, so I'll change it.

I also don't mind to stop returning value if you think it is better.